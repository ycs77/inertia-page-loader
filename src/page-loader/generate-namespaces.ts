import fs from 'fs'
import path from 'path'
import type { UnpluginContextMeta } from 'unplugin'
import type { ResolvedOptions } from '../types'
import { resolveNamespaces } from './namespace-option'

export function generateNamespacesCode(options: ResolvedOptions, meta: UnpluginContextMeta) {
  const cwd = options.cwd
  const appPath = getAppPath(options.appPath)
  const namespaces = resolveNamespaces(cwd, options.namespaces)

  let namespacesCode = Object.keys(namespaces).map(namespace => {
    const modules = namespaces[namespace]

    let code = `\n        '${namespace}': [`
    if (meta.framework === 'vite') {
      modules.forEach(mod => {
        code += `\n          () => resolvePageWithVite(page, import.meta.${!options.ssr ? 'glob' : 'globEager'}('/${resolvePagesDir(cwd, appPath, mod, false)}/**/*${options.extension}'), false),`
      })
    } else {
      modules.forEach(mod => {
        code += `\n          () => require(\`${resolvePagesDir(cwd, appPath, mod)}/\${page}${options.extension}\`),`
      })
    }
    code += '\n        ],'

    return code
  }).join('')

  namespacesCode += '\n      '

  return `{${namespacesCode}}`
}

export function getAppPath(appPath?: string) {
  const resolveAppPath = (appPath: string) => {
    const appFullPath = path.resolve(process.cwd(), appPath).replaceAll('\\', '/')
    if (fs.existsSync(appFullPath)) {
      return appFullPath
    }
  }

  if (appPath) {
    const appFullPath = resolveAppPath(appPath)
    if (appFullPath) return appFullPath
    throw new Error(`[inertia-plugin]: App file "${appFullPath}" does not exist`)
  }

  for (const appPath of [
    'resources/js/app.js',
    'resources/js/app.ts',
    'resources/js/main.js',
    'resources/js/main.ts',
    'src/app.js',
    'src/app.ts',
    'src/main.js',
    'src/main.ts',
    'app.js',
    'app.ts',
    'main.js',
    'main.ts',
  ]) {
    const appFullPath = resolveAppPath(appPath)
    if (appFullPath) return appFullPath
  }
  throw new Error('[inertia-plugin]: App file does not exist')
}

export function resolvePagesDir(cwd: string, appPath: string, dir: string, prefix = true) {
  const rootDir = path.dirname(appPath)
  const pagesDir = path.resolve(
    cwd,
    dir.replace(/^\.\//, '').replace(/\/$/, '')
  )

  if (!fs.existsSync(pagesDir)) {
    throw new Error(`[inertia-plugin]: Pages directory "${pagesDir}" does not exists`)
  }

  const relativePath = path
    .relative(rootDir, pagesDir)
    .replaceAll('\\', '/')

  const relativePrefix = prefix
    ? (/^\.?\.\//.test(relativePath) ? '' : './')
    : ''

  return relativePrefix + relativePath
}
