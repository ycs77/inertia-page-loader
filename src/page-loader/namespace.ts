import fs from 'fs'
import path from 'path'
import type { Namespaces } from '../types'

interface PackageNamespaceExtractorOptions {
  name: string
  filename: string
  dir: string
  parse(content: string): Record<string, string | string[]>
}

function createPackageNamespaceExtractor(options: PackageNamespaceExtractorOptions) {
  return (pkg: string, dir: string = options.dir) => {
    const fullpath = path.resolve(process.cwd(), dir, pkg, options.filename)
    if (!fs.existsSync(fullpath)) {
      throw new Error(`[inertia-plugin]: The ${options.name} "${pkg}" does not exist`)
    }

    const namespaces = options.parse(fs.readFileSync(fullpath, { encoding: 'utf-8' }))
    if (!namespaces) {
      throw new Error(`[inertia-plugin]: The ${options.filename} parse error of "${pkg}"`)
    }

    for (const namespace in namespaces) {
      const mod = namespaces[namespace]
      if (Array.isArray(mod)) {
        namespaces[namespace] = mod.map(mod => path.resolve(dir, pkg, mod))
      } else {
        namespaces[namespace] = path.resolve(dir, pkg, mod)
      }
    }
    return namespaces
  }
}

const npm = createPackageNamespaceExtractor({
  name: 'NPM package',
  filename: 'package.json',
  dir: 'node_modules',
  parse(content) {
    return JSON.parse(content).inertia
  },
})

const composer = createPackageNamespaceExtractor({
  name: 'Composer package',
  filename: 'composer.json',
  dir: 'vendor',
  parse(content) {
    return JSON.parse(content).extra.inertia
  },
})

export function resolveNamespaces(namespaces: Namespaces) {
  const resolvedNamespaces = typeof namespaces === 'function'
    ? namespaces({ npm, composer })
    : namespaces

  const output = {} as Record<string, string[]>

  for (const obj of resolvedNamespaces) {
    for (const key of Object.keys(obj)) {
      output[key] = Array.isArray(obj[key])
        ? (obj[key] as string[])
        : ([obj[key]] as string[])
    }
  }

  return output
}
