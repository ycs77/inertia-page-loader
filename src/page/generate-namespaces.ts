import type { UnpluginContextMeta } from 'unplugin'
import type { ResolvedOptions } from '../types'
import { resolveNamespaces } from './namespace-option'

export function generateNamespacesCode(options: ResolvedOptions, meta: UnpluginContextMeta) {
  const cwd = options.cwd
  const namespaces = resolveNamespaces(cwd, options.namespaces)

  let namespacesCode = Object.keys(namespaces).map(namespace => {
    const modules = namespaces[namespace]

    let code = `\n        '${namespace}': [`
    if (meta.framework === 'vite') {
      modules.forEach(moduleDir => {
        code += `\n          name => resolveVitePage(name, import.meta.${!options.ssr ? 'glob' : 'globEager'}('/${moduleDir}/**/*${options.extension}'), false),`
      })
    } else if (meta.framework === 'webpack') {
      modules.forEach(moduleDir => {
        const moduleImporter = options.ssr || (!options.ssr && !options.import) ? 'require' : 'import'
        code += `\n          name => ${moduleImporter}(\`../${moduleDir}/\${name}${options.extension}\`),`
      })
    }
    code += '\n        ],'

    return code
  }).join('')

  namespacesCode += '\n      '

  return `{${namespacesCode}}`
}
