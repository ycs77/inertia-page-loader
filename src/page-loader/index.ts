import type { UnpluginContextMeta } from 'unplugin'
import type { ResolvedOptions } from '../types'
import { generateNamespacesCode } from './generate-namespaces'

export function pageLoader(options: ResolvedOptions, meta: UnpluginContextMeta) {
  const namespacesCode = generateNamespacesCode(options, meta)

  return `
export function resolvePage(resolver) {
  return async name => await resolvePluginPage(name) ?? await resolver(name)
}

export async function resolvePluginPage(name) {
  if (name.includes('${options.separator}')) {
    const [namespace, page] = name.split('${options.separator}')
    if (namespace && page) {
      const namespaces = ${namespacesCode}
      if (!namespaces[namespace]) {
        throw new Error(\`[inertia-plugin]: Namespace "\${namespace}" not found\`)
      }
      for (const importedNamespace of namespaces[namespace]) {
        if (importedNamespace && typeof importedNamespace === 'function') {
          return importedNamespace()
        }
      }
    }
  }
}

export async function resolvePageWithVite(name, pages, throwNotFoundError = true) {
  for (const path in pages) {
    if (path.endsWith(\`\${name.replace('.', '/')}${options.extension}\`)) {
      const module = typeof pages[path] === 'function'
        ? pages[path]()
        : pages[path]

      return await Promise.resolve(module).then(module => module.default || module)
    }
  }

  if (throwNotFoundError) {
    throw new Error(\`[inertia-plugin]: Page "\${name}" not found\`)
  }
}`
}
