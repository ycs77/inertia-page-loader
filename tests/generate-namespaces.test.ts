import path from 'path'
import { describe, expect, it } from 'vitest'
import { generateNamespacesCode } from '../src/page-loader/generate-namespaces'
import type { Namespaces, ResolvedOptions } from '../src/types'

describe('generate namespaces', () => {
  const baseOptions = <ResolvedOptions>{
    cwd: path.resolve(process.cwd(), 'tests'),
    namespaces: [],
    separator: '::',
    extension: '.vue',
    import: false,
    ssr: false,
  }

  const namespaces = <Namespaces>[
    { MyPackage1: 'test_node_modules/my-plugin1/Pages' },
    { MyPackage2: 'test_node_modules/my-plugin2/other-pages' },
    { MyPhpPackage: 'test_vendor/ycs77/my-php-package/resources/js/Pages' },
  ]

  it('simple options', () => {
    const namespacesCode = generateNamespacesCode(baseOptions, { framework: 'vite' })

    expect(namespacesCode).toMatchSnapshot()
  })

  it('namespaces with vite', () => {
    const options = <ResolvedOptions>{
      ...baseOptions,
      namespaces,
    }

    const namespacesCode = generateNamespacesCode(options, { framework: 'vite' })

    expect(namespacesCode).toMatchSnapshot()
  })

  it('vite with ssr', () => {
    const options = <ResolvedOptions>{
      ...baseOptions,
      namespaces,
      ssr: true,
    }

    const namespacesCode = generateNamespacesCode(options, { framework: 'vite' })

    expect(namespacesCode).toMatchSnapshot()
  })

  it('namespaces with webpack', () => {
    const options = <ResolvedOptions>{
      ...baseOptions,
      namespaces,
      extension: '',
    }

    const namespacesCode = generateNamespacesCode(options, { framework: 'webpack' })

    expect(namespacesCode).toMatchSnapshot()
  })

  it('webpack with import', () => {
    const options = <ResolvedOptions>{
      ...baseOptions,
      namespaces,
      extension: '',
      import: true,
    }

    const namespacesCode = generateNamespacesCode(options, { framework: 'webpack' })

    expect(namespacesCode).toMatchSnapshot()
  })
})
