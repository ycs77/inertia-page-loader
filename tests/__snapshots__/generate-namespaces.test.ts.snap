// Vitest Snapshot v1

exports[`generate namespaces > namespaces with vite 1`] = `
"{
        'MyPackage1': [
          () => resolvePageWithVite(page, import.meta.glob('/test_node_modules/my-plugin1/Pages/**/*.vue'), false),
        ],
        'MyPackage2': [
          () => resolvePageWithVite(page, import.meta.glob('/test_node_modules/my-plugin2/other-pages/**/*.vue'), false),
        ],
        'MyPhpPackage': [
          () => resolvePageWithVite(page, import.meta.glob('/test_vendor/ycs77/my-php-package/resources/js/Pages/**/*.vue'), false),
        ],
      }"
`;

exports[`generate namespaces > namespaces with webpack 1`] = `
"{
        'MyPackage1': [
          () => require(\`../test_node_modules/my-plugin1/Pages/\${page}\`),
        ],
        'MyPackage2': [
          () => require(\`../test_node_modules/my-plugin2/other-pages/\${page}\`),
        ],
        'MyPhpPackage': [
          () => require(\`../test_vendor/ycs77/my-php-package/resources/js/Pages/\${page}\`),
        ],
      }"
`;

exports[`generate namespaces > simple options 1`] = `
"{
      }"
`;

exports[`generate namespaces > vite with ssr 1`] = `
"{
        'MyPackage1': [
          () => resolvePageWithVite(page, import.meta.globEager('/test_node_modules/my-plugin1/Pages/**/*.vue'), false),
        ],
        'MyPackage2': [
          () => resolvePageWithVite(page, import.meta.globEager('/test_node_modules/my-plugin2/other-pages/**/*.vue'), false),
        ],
        'MyPhpPackage': [
          () => resolvePageWithVite(page, import.meta.globEager('/test_vendor/ycs77/my-php-package/resources/js/Pages/**/*.vue'), false),
        ],
      }"
`;

exports[`generate namespaces > webpack with import 1`] = `
"{
        'MyPackage1': [
          () => import(\`../test_node_modules/my-plugin1/Pages/\${page}\`),
        ],
        'MyPackage2': [
          () => import(\`../test_node_modules/my-plugin2/other-pages/\${page}\`),
        ],
        'MyPhpPackage': [
          () => import(\`../test_vendor/ycs77/my-php-package/resources/js/Pages/\${page}\`),
        ],
      }"
`;
